cogs/Moderation.py

```diff
@@ -1,18 +1,114 @@
+from dataclasses import dataclass
+from typing import TypeAlias
+from hashlib import sha256
+import aiohttp
+import re
+import time
+
 import discord
 from discord.commands import option
-from discord.ext import commands
+from discord.ext import commands, tasks
 
 from util.EmbedBuilder import EmbedBuilder
+from util.Logging import log
+
+# hashlib just returns a bytes object
+Hash: TypeAlias = bytes
+
+# this is a simple structure that stores important information
+@dataclass
+class MessageFingerprint:
+    created_at: float  # unix timestamp
+    author_id: int
+    guild_id: int | None  # will be None for DMs
+    channel_id: int
+    jump_url: str  # just so that we can easily refer to this message when surfacing it to humans
+
+    attachment_urls: list[str]  # the discord content URLs for each of the message's uploaded attachments
+    cached_attachment_hashes: set[Hash] | None = None  # populated on the first call to `get_attachment_hashes`
+
+    # hash of the message body (whitespace and case insensitive), or `None` if there is no message body
+    content_hash: Hash | None = None
+
+    # shortcut to build a fingerprint given a message
+    @classmethod
+    def build(cls, message: discord.Message) -> "MessageFingerprint":
+        content_without_whitespace = re.sub(r"\s", "", message.content)
+        return cls(
+            created_at=time.time(),
+            author_id=message.author.id,
+            guild_id=message.guild.id if message.guild else None,
+            channel_id=message.channel.id,
+            jump_url=message.jump_url,
+            attachment_urls=[attachment.url for attachment in message.attachments],
+            content_hash=cls.hash(content_without_whitespace.casefold()) if content_without_whitespace else None,
+        )
+
+    # performs a SHA256 hash
+    @staticmethod
+    def hash(data: str | bytes) -> Hash:
+        if isinstance(data, str):
+            data = data.encode()
+
+        return sha256(data).digest()
+
+    # retrieves and caches attachment hashes
+    # the first call will actually download every attachment,
+    # and all subsequent calls will just return the cached values
+    async def get_attachment_hashes(self) -> set[Hash]:
```

Comment:
Having a method like this is important because we don't want to eagerly download every attachment that everyone ever sends - that would be quite wasteful. 90% of the time, we won't ever even need to download the attachment since it will be pretty rare for someone to send two attachments within 60 seconds

Reply:
Is downloading attachments really needed? I think 99.9% of multiposters will not change the filename of their image when multiposting. attachment url's are in the following format:
`https://cdn.discordapp.com/attachments/channel_id/attachment_id/filename.extension`
F.ex:
`https://cdn.discordapp.com/attachments/1040641081970270288/1040642065312268389/mojito-1592951385.jpg`

I think it would be better if we only hash the filename after the final `/`, this would save me (the host) a lot of bandwith.

Reply:
@sebastiaan-daniels Mentioned this in discord but I'm a little worried that that would raise false positives since screenshots on Windows are always named `image.png`. This function will only download attachments if the same user sends multiple messages with attachments in under 60 seconds, which should be pretty rare. How would you feel about merging this as-is and reverting it if the bandwidth becomes a problem?

Reply:
We can do that, but I'll monitor the bandwith and if it exceeds past my liking, we'll need to either revert or shorten the timespan.